packages/flutter/lib/src/services/platform_channel.dart

```diff
@@ -3,14 +3,124 @@
+  const _ProfiledBinaryMessenger(this.proxy, this.channelTypeName, this.codecTypeName);
+  final BinaryMessenger proxy;
+  final String channelTypeName;
+  final String codecTypeName;
+
+  @override
+  Future<void> handlePlatformMessage(String channel, ByteData? data, PlatformMessageResponseCallback? callback) {
+    return proxy.handlePlatformMessage(channel, data, callback);
+  }
+
+  Future<ByteData?>? sendWithPostfix(String channel, String postfix, ByteData? message) async {
+    final TimelineTask task = TimelineTask();
+    _debugRecordUpStream(channelTypeName, '$channel$postfix', codecTypeName, message);
+    task.start('Platform Channel send $channel$postfix');
+    final ByteData? result;
+    try {
+      result = await proxy.send(channel, message);
+    } finally {
+      task.finish();
+    }
+    _debugRecordDownStream(channelTypeName, '$channel$postfix', codecTypeName, result);
+    return result;
+  }
+
+  @override
+  Future<ByteData?>? send(String channel, ByteData? message) =>
+    sendWithPostfix(channel, '', message);
+
+  @override
+  void setMessageHandler(String channel, MessageHandler? handler) {
+    proxy.setMessageHandler(channel, handler);
+  }
+}
+
+class _PlatformChannelStats {
+  _PlatformChannelStats(this.channel, this.codec, this.type);
+
+  final String channel;
+  final String codec;
+  final String type;
+
+  int _upCount = 0;
+  int _upBytes = 0;
+  int get upBytes => _upBytes;
+  void addUpStream(int bytes) {
+    _upCount += 1;
+    _upBytes += bytes;
+  }
+
+  int _downCount = 0;
+  int _downBytes = 0;
+  int get downBytes => _downBytes;
+  void addDownStream(int bytes) {
+    _downCount += 1;
+    _downBytes += bytes;
+  }
+
+  double get averageUpPayload => _upBytes / _upCount;
+  double get averageDownPayload => _downBytes / _downCount;
+}
+
+final Map<String, _PlatformChannelStats> _debugProfilePlatformChannelsStats = <String, _PlatformChannelStats>{};
+
+Future<void> _debugLaunchProfilePlatformChannels() async {
+  if (!_debugProfilePlatformChannelsIsRunning) {
+    _debugProfilePlatformChannelsIsRunning = true;
+    await Future<dynamic>.delayed(_debugProfilePlatformChannelsRate);
+    _debugProfilePlatformChannelsIsRunning = false;
+    final StringBuffer log = StringBuffer();
+    log.writeln('Platform Channel Stats:');
+    final List<_PlatformChannelStats> allStats =
+        _debugProfilePlatformChannelsStats.values.toList();
+    // Sort highest combined bandwidth first.
+    allStats.sort((_PlatformChannelStats x, _PlatformChannelStats y) =>
+        (y.upBytes + y.downBytes) - (x.upBytes + x.downBytes));
+    for (final _PlatformChannelStats stats in allStats) {
+      log.writeln(
+          '  (name:"${stats.channel}" type:"${stats.type}" codec:"${stats.codec}" upBytes:${stats.upBytes} upBytes_avg:${stats.averageUpPayload.toStringAsFixed(1)} downBytes:${stats.downBytes} downBytes_avg:${stats.averageDownPayload.toStringAsFixed(1)})');
+    }
+    debugPrint(log.toString());
```

Comment:
@kenzieschmoll is there a better way to do this to be able to surface in devtools?

Reply:
Timeline events are probably the best way to surface this in DevTools, which this PR already accomplishes.

Reply:
@kenzieschmoll I think at some point it might be worth considering surfacing this in DevTools.  So the timeline events are showing the time between sending and receiving a response which is helpful.  The stats getting printed out here are the bandwidth of each channel, roughly bytes/sec up and down.  I'm not sure how we decide what is a good feature to do for devtools but this is potentially helpful for inspecting plugins.  I think we should probably wait to get feedback from `customer: money` before implementing it into devtools.

Reply:
We can add a service extension that will modify `debugProfilePlatformChannels` and expose this as a user-controlled toggle in the DevTools performance page. Maybe we should consider turning this flag on by default in debug mode. 

Are all the stats you mentioned included in the timeline event args?

Reply:
I don't think we want to turn it on by default because we have no idea how much overhead it will create because the overhead is a function of what plugins are being used.

I didn't attach bandwidth information to the timeline events.  I'm not sure if that is helpful compared to seeing the summary for a time range.  We don't want people to get out their calculators and start clicking event my event hehe.

Reply:
Ah I see. If the information printed to console ends up being useful, we could send it to devtools via postEvent, and then we can listen and surface it somewhere. We may want to verify its usefulness first though before we plumb this through. The prints will already show up in the DevTools logging view with the current impl.